pkg/cli/adbtunnel.go

```diff
@@ -0,0 +1,234 @@
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"fmt"
+	"net"
+	"sync/atomic"
+
+	wclient "github.com/google/cloud-android-orchestration/pkg/webrtcclient"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pion/webrtc/v3"
+	"github.com/spf13/cobra"
+)
+
+const hostFlag string = "host"
+
+type adbTunnelFlags struct {
+	*subCommandFlags
+	host string
+}
+
+func newADBTunnelCommand(cfgFlags *configFlags) *cobra.Command {
+	adbTunnelFlags := &adbTunnelFlags{&subCommandFlags{cfgFlags, false}, ""}
+	open := &cobra.Command{
+		Use:   "open",
+		Short: "Opens an ADB tunnel.",
+		RunE: func(c *cobra.Command, args []string) error {
+			return runOpenADBTunnelCommand(adbTunnelFlags, &command{c, &adbTunnelFlags.Verbose}, args)
+		},
+	}
+	list := &cobra.Command{
+		Use:   "list",
+		Short: "Lists open ADB tunnels.",
+		RunE:  notImplementedCommand,
+	}
+	close := &cobra.Command{
+		Use:   "close <foo> <bar> <baz>",
+		Short: "Close ADB tunnels.",
+		RunE:  notImplementedCommand,
+	}
+	adbTunnel := &cobra.Command{
+		Use:   "adbtunnel",
+		Short: "Work with ADB tunnels",
+	}
+	addCommonSubcommandFlags(adbTunnel, adbTunnelFlags.subCommandFlags)
+	adbTunnel.PersistentFlags().StringVar(&adbTunnelFlags.host, hostFlag, "", "Specifies the host")
+	adbTunnel.AddCommand(open)
+	adbTunnel.AddCommand(list)
+	adbTunnel.AddCommand(close)
+	return adbTunnel
+}
+
+func runOpenADBTunnelCommand(flags *adbTunnelFlags, c *command, args []string) error {
+	apiClient, err := buildAPIClient(flags.subCommandFlags, c.Command)
+	if err != nil {
+		return err
+	}
+
+	var merr error
+	var conns []*wclient.Connection
+
+	for _, device := range args {
+		observer := ADBForwarder{
+			cmd:    c,
+			host:   flags.host,
+			device: device,
+		}
+		conn, err := apiClient.ConnectWebRTC(flags.host, device, &observer)
+		if err != nil {
+			err = fmt.Errorf("ADB tunnel creation failed for %s on %s: %w", device, flags.host, err)
+			c.PrintErrln(err)
+			merr = multierror.Append(merr, err)
+		} else {
+			// TODO(jemoreira): close everything except the ADB data channel.
+			conns = append(conns, conn)
+		}
+	}
+
+	if len(conns) == 0 {
+		// Return if no tunnels were successfully created
+		return merr
+	}
+
+	// Wait forever
+	select {}
+}
+
+// Forwards ADB messages between a local ADB server and a remote CVD.
+// Implements the Observer interface for the webrtc client.
+type ADBForwarder struct {
+	cmd      *command
+	dc       *webrtc.DataChannel
+	host     string
+	device   string
+	listener net.Listener
+	conn     net.Conn
+	port     int
+	running  atomic.Bool
+}
```

Comment:
nit: wondering whether it's better to isolate the immutable properties like `cmd`, `host` and `device` into a sub structure, so it's more clear what changes and what not.

Reply:
I don't like having those 3 in the ADBForwarder either, the only reason is so I can print logs. But I don't think the extra structure will help readability.

Grouped them together and added an empty line to separate them instead.